src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -4188,23 +4117,38 @@ def _array_syntax_to_indexed(self, parent, loop_vars):
         :raises NotImplementedError: if array sections of differing ranks are
                                      found.
         '''
-        assigns = parent.walk(Assignment)
-        # Check that the LHS of any assignment uses array notation.
-        # Note that this will prevent any WHERE blocks that contain scalar
-        # assignments from being handled but is a necessary limitation until
-        # #717 is done and we interrogate the type of each symbol.
-        for assign in assigns:
-            _ = self._array_notation_rank(assign.lhs)
-
-        # TODO #717 if the supplied code accidentally omits array
+        # If the supplied code accidentally omits array
         # notation for an array reference on the RHS then we will
-        # identify it as a scalar and the code produced from the
-        # PSyIR (using e.g. the Fortran backend) will not
-        # compile. We need to implement robust identification of the
-        # types of all symbols in the PSyIR fragment.
+        # identify it as a scalar and the produce array notated code
+        # in the PSyIR if possible.
         table = parent.scope.symbol_table
         one = Literal("1", INTEGER_TYPE)
+        refs = parent.walk(Reference)
         arrays = parent.walk(ArrayMixin)
+        for ref in refs:
+            add_op = BinaryOperation.Operator.ADD
+            sub_op = BinaryOperation.Operator.SUB
+            if (type(ref) is Reference and type(ref.symbol) is not Symbol and
+                    isinstance(ref.symbol.datatype, ArrayType)):
+                # Skip if the shape doesn't match the loop_vars length.
+                if len(ref.datatype.shape) != len(loop_vars):
+                    continue
+                # Have an implicit full range reference to an array which
```

Comment:
Could you use `Reference2ArrayRangeTrans` here?

Reply:
In fact, you might be better off running over the whole tree and applying this like in utils.py:
https://github.com/stfc/PSyclone/blob/d3f9715900c8cf655d0892fb4ddf4ec29f3f34ed/examples/nemo/scripts/utils.py#L186-L196

Reply:
I probably can, I'll try to give that a go now I understand a bit how the WHERE thing is working.

Reply:
Its looking like the transformation can basically do everything we need for the non-structure cases here (which is what we want) so I am hoping I can just apply that generally before handling the where clauses.

Reply:
There is a case where if we had `WHERE (arr1(1:20) < arr2)` this will currently fail and just generate us a codeblock still - but I'm not actually certain what the original Fortran means/if it makes sense.